lofire-repo/src/tree.rs

```diff
@@ -1 +1,692 @@
+            }
+            Ok(())
+        }
+
+        let mut leaves: Vec<u8> = vec![];
+        let parents = vec![(self.root_id, self.root_key.unwrap())];
+        match collect_leaves(&mut leaves, &self.nodes, &parents, self.nodes.len() - 1) {
+            Ok(_) => Ok(leaves),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use crate::store::*;
+    use crate::tree::*;
+    use crate::types::*;
+    use lofire::types::*;
+
+    // Those constants are calculated with Store::get_max_value_size
+
+    /// Maximum arity of branch containing max number of leaves
+    const MAX_ARITY_LEAVES: usize = 31774;
+    /// Maximum arity of root branch
+    const MAX_ARITY_ROOT: usize = 31770;
+    /// Maximum data that can fit in object.content
+    const MAX_DATA_PAYLOAD_SIZE: usize = 2097112;
+
+    #[test]
+    pub fn test_tree() {
+        let content: Vec<u8> = vec![100; 20 * 4096];
+        //Vec::from("AAABBBCCCDDDEEEFFFGGGHHHIIIJJJKKKLLLMMMNNNOOOPPPQQQRRRSSSTTT");
+        let deps: Vec<ObjectId> = vec![Digest::Blake3Digest32([9; 32])];
+        let expiry = Some(2342);
+        let max_object_size = 0;
+
+        let repo_secret = SymKey::ChaCha20Key([0; 32]);
+        let repo_pubkey = PubKey::Curve25519PubKey([1; 32]);
+
+        let tree = Tree::new(
+            content.clone(),
+            deps,
+            expiry,
+            max_object_size,
+            repo_pubkey,
+            repo_secret,
+        );
+
+        println!("root_id: {:?}", tree.root_id());
+        println!("root_key: {:?}", tree.root_key().unwrap());
+        println!("nodes.len: {:?}", tree.nodes().len());
+        println!("nodes: {:?}", tree.nodes());
+        let mut i = 0;
+        for node in tree.nodes() {
+            println!("#{}: {:?}", i, Tree::object_id(node));
+            i += 1;
+        }
+
+        match tree.content() {
+            Ok(cnt) => {
+                assert_eq!(content, cnt);
+            }
+            Err(e) => panic!("Tree parse error: {:?}", e),
+        }
+
+        let root = tempfile::Builder::new()
+            .prefix("test-tree")
+            .tempdir()
+            .unwrap();
+        let key: [u8; 32] = [0; 32];
+        std::fs::create_dir_all(root.path()).unwrap();
+        println!("{}", root.path().to_str().unwrap());
+        let store = Store::open(root.path(), key);
+
+        tree.save(&store);
+
+        let tree2 = Tree::load(&store, tree.root_id(), tree.root_key()).unwrap();
+
+        println!("nodes2.len: {:?}", tree2.nodes().len());
+        println!("nodes2: {:?}", tree2.nodes());
+        let mut i = 0;
+        for node in tree2.nodes() {
+            println!("#{}: {:?}", i, Tree::object_id(node));
+            i += 1;
+        }
+
+        match tree2.content() {
+            Ok(cnt) => {
+                assert_eq!(content, cnt);
+            }
+            Err(e) => panic!("Tree2 parse error: {:?}", e),
+        }
+    }
+
+    #[test]
+    pub fn test_depth_1() {
```

Comment:
i added this test (test_depth_1) and should assert that a self contained object (with a size of content smaller than what the object can take) does not create a tree, but in fact, just one object that is root and leaf at the same time.
the test does not pass.